scipy/sparse/linalg/_interface.py

```diff
@@ -175,11 +183,29 @@ def __init__(self, dtype, shape):
         self.shape = shape
 
     def _init_dtype(self):
-        """Called from subclasses at the end of the __init__ routine.
+        """Determine the dtype by executing `matvec` on an `int8` test vector.
+
+        In `np.promote_types` hierarchy, the type `int8` is the smallest,
+        so we call `matvec` on `int8` and use the promoted dtype of the output
+        to set the default `dtype` of the `LinearOperator`.
+        We assume that `matmat`, `rmatvec`, and `rmatmat` would result in
+        the same dtype of the output given an `int8` input as `matvec`.
+
+        Called from subclasses at the end of the __init__ routine.
         """
         if self.dtype is None:
-            v = np.zeros(self.shape[-1])
-            self.dtype = np.asarray(self.matvec(v)).dtype
+            v = np.zeros(self.shape[-1], dtype=np.int8)
+            try:
+                matvec_v = np.asarray(self.matvec(v))
+            except OverflowError:
+                # Python large `int` promoted to `np.int64`or `np.int32`
+                try:
+                    test_int64 = np.int64(1)  # noqa: F841
+                    self.dtype = np.int64
+                except AttributeError:
+                    self.dtype = np.int32
```

Comment:
```suggestion
                # Python large `int` promoted to `np.int64`or `np.int32`
                self.dtype = np.dtype(int)
```

Reply:
Thanks very much! This seems to work, with the failure in the benchmark that appears unrelated.